File: src/main/kotlin/com/huawei/cangjie/descriptors/impl/VariableDescriptorImpl.java
package com.huawei.cangjie.descriptors.impl;

import com.huawei.cangjie.descriptors.*;
import com.huawei.cangjie.descriptors.annotations.Annotations;
import com.huawei.cangjie.name.Name;
import com.huawei.cangjie.resolve.scopes.receivers.ContextReceiver;
import com.huawei.cangjie.resolve.scopes.receivers.ExtensionReceiver;
import com.huawei.cangjie.resolve.scopes.receivers.ImplicitContextReceiver;
import com.huawei.cangjie.types.*;
import com.huawei.cangjie.utils.ReadOnly;
import com.huawei.cangjie.utils.SmartSet;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.*;


public class VariableDescriptorImpl extends VariableDescriptorWithInitializerImpl implements VariableDescriptor {
    private final Modality modality;
    private final VariableDescriptor original;
    private final CallableMemberDescriptor.Kind kind;

    private DescriptorVisibility visibility;
    private Collection<? extends  VariableDescriptor> overriddenProperties = null;
    private List<ReceiverParameterDescriptor> contextReceiverParameters = Collections.emptyList();
    private ReceiverParameterDescriptor dispatchReceiverParameter;
    private ReceiverParameterDescriptor extensionReceiverParameter;
    private List<TypeParameterDescriptor> typeParameters;
    //    private PropertyGetterDescriptorImpl getter;
//    private PropertySetterDescriptor setter;

//    private FieldDescriptor backingField;
//    private FieldDescriptor delegateField;

    protected VariableDescriptorImpl(
            @NotNull DeclarationDescriptor containingDeclaration,
            @Nullable VariableDescriptor original,
            @NotNull Annotations annotations,
            @NotNull Modality modality,
            @NotNull DescriptorVisibility visibility,
            boolean isVar,
            @NotNull Name name,
            @NotNull CallableMemberDescriptor.Kind kind,
            @Nullable CangJieType outType,
            @NotNull SourceElement source


    ) {
        super(containingDeclaration, annotations, name, outType, isVar, source);
        this.modality = modality;
        this.visibility = visibility;
        this.original = original == null ? this : original;
        this.kind = kind;
//        this.lateInit = lateInit;
//        this.isConst = isConst;
//        this.isExpect = isExpect;
//        this.isActual = isActual;
//        this.isExternal = isExternal;

    }
    protected VariableDescriptorImpl(
            @NotNull DeclarationDescriptor containingDeclaration,
            @Nullable VariableDescriptor original,
            @NotNull Annotations annotations,
            @NotNull Modality modality,
            @NotNull DescriptorVisibility visibility,
            boolean isVar,
            @NotNull Name name,
            @NotNull CallableMemberDescriptor.Kind kind,
            @NotNull SourceElement source


    ) {
        super(containingDeclaration, annotations, name, null, isVar, source);
        this.modality = modality;
        this.visibility = visibility;
        this.original = original == null ? this : original;
        this.kind = kind;
//        this.lateInit = lateInit;
//        this.isConst = isConst;
//        this.isExpect = isExpect;
//        this.isActual = isActual;
//        this.isExternal = isExternal;

    }
    @NotNull
    public static VariableDescriptorImpl create(
            @NotNull DeclarationDescriptor containingDeclaration,
            @NotNull Annotations annotations,
            @NotNull Modality modality,
            @NotNull DescriptorVisibility visibility,
            boolean isVar,
            @NotNull Name name,
            @NotNull CallableMemberDescriptor.Kind kind,
            @NotNull SourceElement source
//            bool lateInit,
//            bool isConst,
//            bool isExpect,
//            bool isActual,
//            bool isExternal,
//            boolean isDelegated
    ) {
        return new VariableDescriptorImpl(containingDeclaration, null, annotations,
                modality, visibility, isVar, name, kind, source
//                , lateInit, isConst,
//                isExpect, isActual, isExternal

        );
    }

//    private static DescriptorVisibility normalizeVisibility(DescriptorVisibility prev, Kind kind) {
//        if (kind == Kind.FAKE_OVERRIDE && DescriptorVisibilities.isPrivate(prev.normalize())) {
//            return DescriptorVisibilities.INVISIBLE_FAKE;
//        }
//        return prev;
//    }

    private static ReceiverParameterDescriptor substituteParameterDescriptor(
            TypeSubstitutor substitutor,
            VariableDescriptor substitutedVariableDescriptor,
            ReceiverParameterDescriptor receiverParameterDescriptor
    ) {
        CangJieType substitutedType = substitutor.substitute(receiverParameterDescriptor.getType(), Variance.IN_VARIANCE);
        if (substitutedType == null) return null;
        return new ReceiverParameterDescriptorImpl(
                substitutedVariableDescriptor,
                new ExtensionReceiver(substitutedVariableDescriptor, substitutedType, receiverParameterDescriptor.getValue()),
                receiverParameterDescriptor.getAnnotations()
        );
    }

    private static ReceiverParameterDescriptor substituteContextParameterDescriptor(
            TypeSubstitutor substitutor,
            VariableDescriptor substitutedVariableDescriptor,
            ReceiverParameterDescriptor receiverParameterDescriptor
    ) {
        CangJieType substitutedType = substitutor.substitute(receiverParameterDescriptor.getType(), Variance.IN_VARIANCE);
        if (substitutedType == null) return null;
        return new ReceiverParameterDescriptorImpl(
                substitutedVariableDescriptor,
                new ContextReceiver(substitutedVariableDescriptor,
                        substitutedType,
                        ((ImplicitContextReceiver) receiverParameterDescriptor.getValue()).getCustomLabelName(),
                        receiverParameterDescriptor.getValue()),
                receiverParameterDescriptor.getAnnotations()
        );
    }

//    private static FunctionDescriptor getSubstitutedInitialSignatureDescriptor(
//            @NotNull TypeSubstitutor substitutor,
//            @NotNull PropertyAccessorDescriptor accessorDescriptor
//    ) {
//        return accessorDescriptor.getInitialSignatureDescriptor() != null
//                ? accessorDescriptor.getInitialSignatureDescriptor().substitute(substitutor)
//                : null;
//    }


    public void setInType(@NotNull CangJieType inType) {
        /* Do nothing as the corresponding setter is generated by default */
    }

    @kotlin.Deprecated(message = "This method is left for binary compatibility with android.nav.safearg plugin. Used in SafeArgSyntheticDescriptorGenerator.kt")
    public void setType(
            @NotNull CangJieType outType,
            @ReadOnly @NotNull List<? extends TypeParameterDescriptor> typeParameters,
            @Nullable ReceiverParameterDescriptor dispatchReceiverParameter,
            @Nullable ReceiverParameterDescriptor extensionReceiverParameter
    ) {
        setType(outType, typeParameters, dispatchReceiverParameter, extensionReceiverParameter,
                Collections.emptyList());
    }

    public void setType(
            @NotNull CangJieType outType,
            @ReadOnly @NotNull List<? extends TypeParameterDescriptor> typeParameters,
            @Nullable ReceiverParameterDescriptor dispatchReceiverParameter,
            @Nullable ReceiverParameterDescriptor extensionReceiverParameter,
            @NotNull List<ReceiverParameterDescriptor> contextReceiverParameters
    ) {
        setOutType(outType);

        this.typeParameters = new ArrayList<TypeParameterDescriptor>(typeParameters);

        this.extensionReceiverParameter = extensionReceiverParameter;
        this.dispatchReceiverParameter = dispatchReceiverParameter;
        this.contextReceiverParameters = contextReceiverParameters;
    }

//    public void initialize(
//            @Nullable PropertyGetterDescriptorImpl getter,
//            @Nullable PropertySetterDescriptor setter
//    ) {
//        initialize(getter, setter, null, null);
//    }

//    public void initialize(
//            @Nullable PropertyGetterDescriptorImpl getter,
//            @Nullable PropertySetterDescriptor setter,
//            @Nullable FieldDescriptor backingField,
//            @Nullable FieldDescriptor delegateField
//    ) {
//        this.getter = getter;
//        this.setter = setter;
//        this.backingField = backingField;
//        this.delegateField = delegateField;
//    }

    @NotNull
    @Override
    public List<TypeParameterDescriptor> getTypeParameters() {
        List<TypeParameterDescriptor> parameters = typeParameters;
        // Diagnostics for EA-212070
        if (parameters == null) {
            throw new IllegalStateException("typeParameters == null for " + this);
        }
        return parameters;
    }

    @Override
    public void validate() {
        getTypeParameters();
    }

    @Override
    @NotNull
    public List<ReceiverParameterDescriptor> getContextReceiverParameters() {
        return contextReceiverParameters;
    }

    @Override
    @Nullable
    public ReceiverParameterDescriptor getExtensionReceiverParameter() {
        return extensionReceiverParameter;
    }

    @Nullable
    @Override
    public ReceiverParameterDescriptor getDispatchReceiverParameter() {
        return dispatchReceiverParameter;
    }

    @NotNull
    @Override
    public CangJieType getReturnType() {
        return getType();
    }

    @NotNull
    @Override
    public Modality getModality() {
        return modality;
    }

    @NotNull
    @Override
    public DescriptorVisibility getVisibility() {
        return visibility;
    }

    public void setVisibility(@NotNull DescriptorVisibility visibility) {
        this.visibility = visibility;
    }

//    @Override
//    @Nullable
//    public PropertyGetterDescriptorImpl getGetter() {
//        return getter;
//    }

//    @Override
//    @Nullable
//    public PropertySetterDescriptor getSetter() {
//        return setter;
//    }


    //    @Override
//    public bool isLateInit() {
//        return lateInit;
//    }
//
//    @Override
//    public bool isConst() {
//        return isConst;
//    }
//
//    @Override
//    public bool isExternal() {
//        return isExternal;
//    }
//
//    @Override
//    public boolean isDelegated() {
//        return isDelegated;
//    }

//    @Override
//    @NotNull
//    public List<PropertyAccessorDescriptor> getAccessors() {
//        List<PropertyAccessorDescriptor> result = new ArrayList<PropertyAccessorDescriptor>(2);
//        if (getter != null) {
//            result.add(getter);
//        }
//        if (setter != null) {
//            result.add(setter);
//        }
//        return result;
//    }

    @Override
    public VariableDescriptor substitute(@NotNull TypeSubstitutor originalSubstitutor) {
        if (originalSubstitutor.isEmpty()) {
            return this;
        }

        return Objects.requireNonNull(newCopyBuilder()
                .setSubstitution(originalSubstitutor.getSubstitution())
                .setOriginal(getOriginal())
                .build());
    }

    @NotNull
    @Override
    public CopyConfiguration newCopyBuilder() {
        return new CopyConfiguration();
    }

    @NotNull
    private SourceElement getSourceToUseForCopy(boolean preserveSource, @Nullable VariableDescriptor original) {
        return preserveSource
                ? (original != null ? original : getOriginal()).getSource()
                : SourceElement.NO_SOURCE;
    }

    @Nullable
    protected VariableDescriptor doSubstitute(@NotNull CopyConfiguration copyConfiguration) {
        VariableDescriptorImpl substitutedDescriptor = createSubstitutedCopy(
                copyConfiguration.owner, copyConfiguration.modality, copyConfiguration.visibility,
                copyConfiguration.original, copyConfiguration.kind, copyConfiguration.name,
                getSourceToUseForCopy(copyConfiguration.preserveSourceElement, copyConfiguration.original));

        List<TypeParameterDescriptor> originalTypeParameters =
                copyConfiguration.newTypeParameters == null ? getTypeParameters() : copyConfiguration.newTypeParameters;
        List<TypeParameterDescriptor> substitutedTypeParameters = new ArrayList<TypeParameterDescriptor>(originalTypeParameters.size());
        TypeSubstitutor substitutor = DescriptorSubstitutor.substituteTypeParameters(
                originalTypeParameters, copyConfiguration.substitution, substitutedDescriptor, substitutedTypeParameters
        );

        CangJieType originalOutType = copyConfiguration.returnType;
        CangJieType outType = substitutor.substitute(originalOutType, Variance.OUT_VARIANCE);
        if (outType == null) {
            return null; // TODO : tell the user that the property was projected out
        }

        CangJieType inType = substitutor.substitute(originalOutType, Variance.IN_VARIANCE);

        if (inType != null) {
            substitutedDescriptor.setInType(inType);
        }

        ReceiverParameterDescriptor substitutedDispatchReceiver;
        ReceiverParameterDescriptor dispatchReceiver = copyConfiguration.dispatchReceiverParameter;
        if (dispatchReceiver != null) {
            substitutedDispatchReceiver = dispatchReceiver.substitute(substitutor);
            if (substitutedDispatchReceiver == null) return null;
        } else {
            substitutedDispatchReceiver = null;
        }

        ReceiverParameterDescriptor substitutedExtensionReceiver;
        if (extensionReceiverParameter != null) {
            substitutedExtensionReceiver = substituteParameterDescriptor(substitutor, substitutedDescriptor, extensionReceiverParameter);
        } else {
            substitutedExtensionReceiver = null;
        }

        List<ReceiverParameterDescriptor> substitutedContextReceivers = new ArrayList<ReceiverParameterDescriptor>();
        for (ReceiverParameterDescriptor contextReceiverParameter : contextReceiverParameters) {
            ReceiverParameterDescriptor substitutedContextReceiver = substituteContextParameterDescriptor(substitutor, substitutedDescriptor,
                    contextReceiverParameter);
            if (substitutedContextReceiver != null) {
                substitutedContextReceivers.add(substitutedContextReceiver);
            }
        }

        substitutedDescriptor.setType(outType, substitutedTypeParameters, substitutedDispatchReceiver, substitutedExtensionReceiver,
                substitutedContextReceivers);
//
//        PropertyGetterDescriptorImpl newGetter = getter == null ? null : new PropertyGetterDescriptorImpl(
//                substitutedDescriptor, getter.getAnnotations(), copyConfiguration.modality, normalizeVisibility(getter.getVisibility(), copyConfiguration.kind),
//                getter.isDefault(), getter.isExternal(), getter.isInline(), copyConfiguration.kind,
//                copyConfiguration.getOriginalGetter(),
//                SourceElement.NO_SOURCE
//        );
//        if (newGetter != null) {
//            CangJieType returnType = getter.getReturnType();
//            newGetter.setInitialSignatureDescriptor(getSubstitutedInitialSignatureDescriptor(substitutor, getter));
//            newGetter.initialize(returnType != null ? substitutor.substitute(returnType, Variance.OUT_VARIANCE) : null);
//        }
//        PropertySetterDescriptorImpl newSetter = setter == null ? null : new PropertySetterDescriptorImpl(
//                substitutedDescriptor, setter.getAnnotations(), copyConfiguration.modality, normalizeVisibility(setter.getVisibility(), copyConfiguration.kind),
//                setter.isDefault(), setter.isExternal(), setter.isInline(), copyConfiguration.kind,
//                copyConfiguration.getOriginalSetter(),
//                SourceElement.NO_SOURCE
//        );
//        if (newSetter != null) {
//            List<ValueParameterDescriptor> substitutedValueParameters = FunctionDescriptorImpl.getSubstitutedValueParameters(
//                    newSetter, setter.getValueParameters(), substitutor, /* dropOriginal = */ false,
//                    false, null
//            );
//            if (substitutedValueParameters == null) {
//                // The setter is projected out, e.g. in this case:
//                //     trait Tr<T> { var v: T }
//                //     fun test(tr: Tr<out Any?>) { ... }
//                // we want to tell the user that although the property is declared as a var,
//                // it can not be assigned to because of the projection
//                substitutedDescriptor.setSetterProjectedOut(true);
//                substitutedValueParameters = Collections.<ValueParameterDescriptor>singletonList(
//                        PropertySetterDescriptorImpl.createSetterParameter(
//                                newSetter,
//                                getBuiltIns(copyConfiguration.owner).getNothingType(),
//                                setter.getValueParameters().get(0).getAnnotations()
//                        )
//                );
//            }
//            if (substitutedValueParameters.size() != 1) {
//                throw new IllegalStateException();
//            }
//            newSetter.setInitialSignatureDescriptor(getSubstitutedInitialSignatureDescriptor(substitutor, setter));
//            newSetter.initialize(substitutedValueParameters.get(0));
//        }

//        substitutedDescriptor.initialize(
//                newGetter,
//                newSetter,
//                backingField == null ? null : new FieldDescriptorImpl(backingField.getAnnotations(), substitutedDescriptor),
//                delegateField == null ? null : new FieldDescriptorImpl(delegateField.getAnnotations(), substitutedDescriptor)
//        );

        if (copyConfiguration.copyOverrides) {
            Collection<CallableMemberDescriptor> overridden = SmartSet.create();
            for (VariableDescriptor propertyDescriptor : getOverriddenDescriptors()) {
                overridden.add(propertyDescriptor.substitute(substitutor));
            }
            substitutedDescriptor.setOverriddenDescriptors(overridden);
        }

//        if (isConst() && compileTimeInitializerFactory != null) {
//            substitutedDescriptor.setCompileTimeInitializer(compileTimeInitializer, compileTimeInitializerFactory);
//        }

        return substitutedDescriptor;
    }

    @NotNull
    protected VariableDescriptorImpl createSubstitutedCopy(
            @NotNull DeclarationDescriptor newOwner,
            @NotNull Modality newModality,
            @NotNull DescriptorVisibility newVisibility,
            @Nullable VariableDescriptor original,
            @NotNull Kind kind,
            @NotNull Name newName,
            @NotNull SourceElement source
    ) {
        return new VariableDescriptorImpl(
                newOwner, original, getAnnotations(), newModality, newVisibility, isVar(), newName, kind, source
//
//                isLateInit(), isConst(), isExpect(), isActual(), isExternal()
//                , isDelegated()
        );
    }

    @Override
    public <R, D> R accept(DeclarationDescriptorVisitor<R, D> visitor, D data) {
        return visitor.visitVariableDescriptor(this, data);
    }

    @NotNull
    @Override
    public VariableDescriptor getOriginal() {
        return original == this ? this : original.getOriginal();
    }

    @NotNull
    @Override
    public CallableMemberDescriptor.Kind getKind() {
        return kind;
    }



//    @Override
//    public bool isExpect() {
//        return isExpect;
//    }
//
//    @Override
//    public bool isActual() {
//        return isActual;
//    }
//
//    @Override
//    @Nullable
//    public FieldDescriptor getBackingField() {
//        return backingField;
//    }
//
//    @Override
//    @Nullable
//    public FieldDescriptor getDelegateField() {
//        return delegateField;
//    }

@Override

    public @NotNull Collection<? extends VariableDescriptor> getOverriddenDescriptors() {
        return overriddenProperties != null ? (Collection<VariableDescriptor>) overriddenProperties : Collections.emptyList();
    }
//


    @Override
    @SuppressWarnings("unchecked")
    public void setOverriddenDescriptors(@NotNull Collection<? extends CallableMemberDescriptor> overriddenDescriptors) {
        this.overriddenProperties = (Collection<? extends VariableDescriptor>) overriddenDescriptors;
    }

    @NotNull
    @Override
    public VariableDescriptor copy(DeclarationDescriptor newOwner, Modality modality, DescriptorVisibility visibility, Kind kind, boolean copyOverrides) {
        //noinspection ConstantConditions
        return newCopyBuilder()
                .setOwner(newOwner)
                .setOriginal(null)
                .setModality(modality)
                .setVisibility(visibility)
                .setKind(kind)
                .setCopyOverrides(copyOverrides)
                .build();
    }


    @Nullable
    @Override
    public <V> V getUserData(UserDataKey<V> key) {
        return null;
    }


    public class CopyConfiguration implements VariableDescriptor.CopyBuilder<VariableDescriptor> {
        private DeclarationDescriptor owner = getContainingDeclaration();
        private Modality modality = getModality();
        private DescriptorVisibility visibility = getVisibility();
        private VariableDescriptor original = null;
        private boolean preserveSourceElement = false;
        private Kind kind = getKind();
        private TypeSubstitution substitution = TypeSubstitution.EMPTY;
        private boolean copyOverrides = true;
        private ReceiverParameterDescriptor dispatchReceiverParameter = VariableDescriptorImpl.this.dispatchReceiverParameter;
        private List<TypeParameterDescriptor> newTypeParameters = null;
        private Name name = getName();
        private CangJieType returnType = getType();

        @NotNull
        @Override
        public CopyConfiguration setOwner(@NotNull DeclarationDescriptor owner) {
            this.owner = owner;
            return this;
        }

        @NotNull
        @Override
        public CopyConfiguration setOriginal(@Nullable CallableMemberDescriptor original) {
            this.original = (VariableDescriptor) original;
            return this;
        }

        @NotNull
        @Override
        public CopyConfiguration setPreserveSourceElement() {
            preserveSourceElement = true;
            return this;
        }

        @NotNull
        @Override
        public CopyBuilder<VariableDescriptor> setReturnType(@NotNull CangJieType type) {
            returnType = type;
            return this;
        }

        @NotNull
        @Override
        public CopyConfiguration setModality(@NotNull Modality modality) {
            this.modality = modality;
            return this;
        }

        @NotNull
        @Override
        public CopyConfiguration setVisibility(@NotNull DescriptorVisibility visibility) {
            this.visibility = visibility;
            return this;
        }

        @NotNull
        @Override
        public CopyConfiguration setKind(@NotNull Kind kind) {
            this.kind = kind;
            return this;
        }

        @NotNull
        @Override
        public CopyBuilder<VariableDescriptor> setTypeParameters(@NotNull List<TypeParameterDescriptor> typeParameters) {
            this.newTypeParameters = typeParameters;
            return this;
        }

        @NotNull
        @Override
        public CopyConfiguration setDispatchReceiverParameter(@Nullable ReceiverParameterDescriptor dispatchReceiverParameter) {
            this.dispatchReceiverParameter = dispatchReceiverParameter;
            return this;
        }

        @NotNull
        @Override
        public CopyConfiguration setSubstitution(@NotNull TypeSubstitution substitution) {
            this.substitution = substitution;
            return this;
        }

        @NotNull
        @Override
        public CopyConfiguration setCopyOverrides(boolean copyOverrides) {
            this.copyOverrides = copyOverrides;
            return this;
        }

        @NotNull
        @Override
        public CopyBuilder<VariableDescriptor> setName(@NotNull Name name) {
            this.name = name;
            return this;
        }

        @Nullable
        @Override
        public VariableDescriptor build() {
            return doSubstitute(this);
        }

//        PropertyGetterDescriptor getOriginalGetter() {
//            if (original == null) return null;
//            return original.getGetter();
//        }
//
//        PropertySetterDescriptor getOriginalSetter() {
//            if (original == null) return null;
//            return original.getSetter();
//        }
    }
}

File: descriptors/src/main/kotlin/org/cangnova/cangjie/descriptors/impl/VariableDescriptorImpl.kt
/*
 * Copyright 2025 LinQingYing. and contributors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * The use of this source code is governed by the Apache License 2.0,
 * which allows users to freely use, modify, and distribute the code,
 * provided they adhere to the terms of the license.
 *
 * The software is provided "as-is", and the authors are not responsible for
 * any damages or issues arising from its use.
 *
 */

package org.cangnova.cangjie.descriptors.impl

import org.cangnova.cangjie.descriptors.*
import org.cangnova.cangjie.descriptors.annotations.Annotations
import org.cangnova.cangjie.name.Name
import org.cangnova.cangjie.psi.CjVariable
import org.cangnova.cangjie.resolve.scopes.receivers.ContextReceiver
import org.cangnova.cangjie.resolve.scopes.receivers.ExtensionReceiver
import org.cangnova.cangjie.resolve.scopes.receivers.ImplicitContextReceiver
import org.cangnova.cangjie.resolve.source.getPsi
import org.cangnova.cangjie.types.*


open class VariableDescriptorImpl(
    containingDeclaration: DeclarationDescriptor,
    name: Name,
    outType: CangJieType?,
    isVar: Boolean,
    source: SourceElement,
    visibility: DescriptorVisibility
) : VariableDescriptorWithInitializerImpl(containingDeclaration, Annotations.EMPTY, name, outType, isVar, source),
    CallableMemberDescriptor {
    override val original: VariableDescriptorImpl
        get() = super.original as VariableDescriptorImpl
    private var _visibility: DescriptorVisibility = visibility


    protected open fun createSubstitutedCopy(
        newOwner: DeclarationDescriptor,
        newModality: Modality,
        newVisibility: DescriptorVisibility,
        original: VariableDescriptor?,
        kind: CallableMemberDescriptor.Kind,

        newName: Name,
        source: SourceElement
    ): VariableDescriptorImpl {
        return VariableDescriptorImpl(
            newOwner, newName, null, isVar, source, newVisibility

        )
    }

    private var _overriddenDescriptors: Collection<VariableDescriptorImpl> = emptyList()



    override val overriddenDescriptors: Collection<VariableDescriptorImpl>
        get() = _overriddenDescriptors

    override val modality: Modality
        get() = Modality.FINAL



    override val kind: CallableMemberDescriptor.Kind
        get() = CallableMemberDescriptor.Kind.DECLARATION

    override fun setOverriddenDescriptors(overriddenDescriptors: Collection<  CallableMemberDescriptor>) {
//        this._overriddenDescriptors = overriddenDescriptors
    }



    override fun <V> getUserData(key: CallableDescriptor.UserDataKey<V>): V? {
        return null
    }
    override fun copy(
        newOwner: DeclarationDescriptor,
        modality: Modality,
        visibility: DescriptorVisibility,
        kind: CallableMemberDescriptor.Kind,
        copyOverrides: Boolean
    ): CallableMemberDescriptor {
        return this
    }

    override fun newCopyBuilder(): CopyConfiguration {
//        return object : CallableMemberDescriptor.CopyBuilder<CallableMemberDescriptor> {
//            override fun setOwner(owner: DeclarationDescriptor): CallableMemberDescriptor.CopyBuilder<CallableMemberDescriptor> {
//                return this
//            }
//
//            override fun setModality(modality: Modality): CallableMemberDescriptor.CopyBuilder<CallableMemberDescriptor> {
//                return this
//
//            }
//
//            override fun setVisibility(visibility: DescriptorVisibility): CallableMemberDescriptor.CopyBuilder<CallableMemberDescriptor> {
//                return this
//
//            }
//
//            override fun setKind(kind: CallableMemberDescriptor.Kind): CallableMemberDescriptor.CopyBuilder<CallableMemberDescriptor> {
//                return this
//
//            }
//
//            override fun setTypeParameters(parameters: MutableList<TypeParameterDescriptor>): CallableMemberDescriptor.CopyBuilder<CallableMemberDescriptor> {
//                return this
//
//            }
//
//            override fun setDispatchReceiverParameter(dispatchReceiverParameter: ReceiverParameterDescriptor?): CallableMemberDescriptor.CopyBuilder<CallableMemberDescriptor> {
//                return this
//
//            }
//
//            override fun setSubstitution(substitution: TypeSubstitution): CallableMemberDescriptor.CopyBuilder<CallableMemberDescriptor> {
//                return this
//
//            }
//
//            override fun setCopyOverrides(copyOverrides: Boolean): CallableMemberDescriptor.CopyBuilder<CallableMemberDescriptor> {
//                return this
//
//            }
//
//            override fun setName(name: Name): CallableMemberDescriptor.CopyBuilder<CallableMemberDescriptor> {
//                return this
//
//            }
//
//            override fun setOriginal(original: CallableMemberDescriptor?): CallableMemberDescriptor.CopyBuilder<CallableMemberDescriptor> {
//                return this
//
//            }
//
//            override fun setPreserveSourceElement(): CallableMemberDescriptor.CopyBuilder<CallableMemberDescriptor> {
//                return this
//
//            }
//
//            override fun setReturnType(type: CangJieType): CallableMemberDescriptor.CopyBuilder<CallableMemberDescriptor> {
//                return this
//
//            }
//
//            override fun build(): CallableMemberDescriptor {
//                return this@VariableDescriptorImpl
//            }
        return CopyConfiguration()

    }


    override fun <R, D> accept(visitor: DeclarationDescriptorVisitor<R, D>, data: D?): R? {
        return visitor.visitVariableDescriptor(this, data)
    }

    override var visibility: DescriptorVisibility = _visibility

    inner class CopyConfiguration : CallableMemberDescriptor.CopyBuilder<VariableDescriptorImpl> {
        var _owner: DeclarationDescriptor = containingDeclaration
        var _modality: Modality = modality
        var _visibility = this@VariableDescriptorImpl.visibility
        var _original: VariableDescriptorImpl? = null
        var _preserveSourceElement = true
        var _kind: CallableMemberDescriptor.Kind = kind
        var _substitution = TypeSubstitution.EMPTY
        var _copyOverrides = true
        var _dispatchReceiverParameter: ReceiverParameterDescriptor? =
            this@VariableDescriptorImpl.dispatchReceiverParameter
        var _newTypeParameters: List<TypeParameterDescriptor>? = null
        var _name = this@VariableDescriptorImpl.name
        var _returnType: CangJieType = type

        override fun setOwner(owner: DeclarationDescriptor): CopyConfiguration {
            this._owner = owner
            return this
        }

        override fun setOriginal(original: CallableMemberDescriptor?): CopyConfiguration {
            this._original = original as VariableDescriptorImpl?
            return this
        }

        override fun setPreserveSourceElement(): CopyConfiguration {
            _preserveSourceElement = true
            return this
        }

        override fun setReturnType(type: CangJieType): CallableMemberDescriptor.CopyBuilder<VariableDescriptorImpl> {
            _returnType = type
            return this
        }

        override fun setModality(modality: Modality): CopyConfiguration {
            this._modality = modality
            return this
        }

        override fun setVisibility(visibility: DescriptorVisibility): CopyConfiguration {
            this._visibility = visibility
            return this
        }

        override fun setKind(kind: CallableMemberDescriptor.Kind): CopyConfiguration {
            this._kind = kind
            return this
        }

        override fun setTypeParameters(parameters: List<TypeParameterDescriptor>): CallableMemberDescriptor.CopyBuilder<VariableDescriptorImpl> {
            this._newTypeParameters = parameters
            return this
        }

        override fun setDispatchReceiverParameter(dispatchReceiverParameter: ReceiverParameterDescriptor?): CopyConfiguration {
            this._dispatchReceiverParameter = dispatchReceiverParameter
            return this
        }

        override fun setSubstitution(substitution: TypeSubstitution): CopyConfiguration {
            this._substitution = substitution
            return this
        }

        override fun setCopyOverrides(copyOverrides: Boolean): CopyConfiguration {
            this._copyOverrides = copyOverrides
            return this
        }

        override fun setName(name: Name): CallableMemberDescriptor.CopyBuilder<VariableDescriptorImpl > {
            this._name = name
            return this
        }

        override fun build(): VariableDescriptorImpl? {
            return doSubstitute(this)
        }


    }

    private fun getSourceToUseForCopy(preserveSource: Boolean, original: VariableDescriptorImpl?): SourceElement {
        return if (preserveSource)
            (original ?: original)!!.source
        else
            SourceElement.NO_SOURCE
    }

    protected fun doSubstitute(copyConfiguration: CopyConfiguration): VariableDescriptorImpl? {
        val substitutedDescriptor = createSubstitutedCopy(
            copyConfiguration._owner, copyConfiguration._modality, copyConfiguration._visibility,
            copyConfiguration._original, copyConfiguration._kind,copyConfiguration._name,
            getSourceToUseForCopy(copyConfiguration._preserveSourceElement, copyConfiguration._original)
        )
        val originalTypeParameters =
            if (copyConfiguration._newTypeParameters == null) typeParameters else copyConfiguration._newTypeParameters
                ?: emptyList()
        val substitutedTypeParameters  = ArrayList<TypeParameterDescriptor>(originalTypeParameters.size)
        val substitutor = DescriptorSubstitutor.substituteTypeParameters(
            originalTypeParameters, copyConfiguration._substitution, substitutedDescriptor, substitutedTypeParameters
        )

        val originalOutType: CangJieType = copyConfiguration._returnType
        val outType = substitutor.substitute(originalOutType, Variance.INVARIANT)
            ?: return null // TODO : tell the user that the property was projected out

        val substitutedDispatchReceiver: ReceiverParameterDescriptor?
        val dispatchReceiver = copyConfiguration._dispatchReceiverParameter
        if (dispatchReceiver != null) {
            substitutedDispatchReceiver = dispatchReceiver.substitute(substitutor)
            if (substitutedDispatchReceiver == null) return null
        } else {
            substitutedDispatchReceiver = null
        }
        val substitutedExtensionReceiver = if (extensionReceiverParameter != null) {
            substituteParameterDescriptor(
                substitutor, substitutedDescriptor,
                extensionReceiverParameter!!
            )
        } else {
            null
        }

        val substitutedContextReceivers: MutableList<ReceiverParameterDescriptor> = ArrayList()
        for (contextReceiverParameter in contextReceiverParameters) {
            val substitutedContextReceiver = substituteContextParameterDescriptor(
                substitutor, substitutedDescriptor,
                contextReceiverParameter
            )
            if (substitutedContextReceiver != null) {
                substitutedContextReceivers.add(substitutedContextReceiver)
            }
        }

        substitutedDescriptor.setType(
            outType, substitutedTypeParameters, substitutedDispatchReceiver, substitutedExtensionReceiver,
            substitutedContextReceivers
        )
        return substitutedDescriptor

    }

    override fun substitute(substitutor: TypeSubstitutor): VariableDescriptorImpl? {
        if (substitutor.isEmpty) {
            return this
        }

        return newCopyBuilder()
            .setSubstitution(substitutor.substitution)
            .setOriginal(original)
            .build()

    }
    companion object {
        private fun substituteContextParameterDescriptor(
            substitutor: TypeSubstitutor,
            substitutedPropertyDescriptor: VariableDescriptorImpl,
            receiverParameterDescriptor: ReceiverParameterDescriptor
        ): ReceiverParameterDescriptor? {
            val substitutedType = substitutor.substitute(receiverParameterDescriptor.type, Variance.INVARIANT)
                ?: return null
            return ReceiverParameterDescriptorImpl(
                substitutedPropertyDescriptor,
                ContextReceiver(
                    substitutedPropertyDescriptor,
                    substitutedType,
                    (receiverParameterDescriptor.value as ImplicitContextReceiver).customLabelName,
                    receiverParameterDescriptor.value
                ),
                receiverParameterDescriptor.annotations
            )
        }

        private fun substituteParameterDescriptor(
            substitutor: TypeSubstitutor,
            substitutedPropertyDescriptor: VariableDescriptor,
            receiverParameterDescriptor: ReceiverParameterDescriptor
        ): ReceiverParameterDescriptor? {
            val substitutedType = substitutor.substitute(receiverParameterDescriptor.type, Variance.INVARIANT)
                ?: return null
            return ReceiverParameterDescriptorImpl(
                substitutedPropertyDescriptor,
                ExtensionReceiver(substitutedPropertyDescriptor, substitutedType, receiverParameterDescriptor.value),
                receiverParameterDescriptor.annotations
            )
        }

        @JvmStatic
        fun create(
            containingDeclaration: DeclarationDescriptor,
            name: Name,
            visibility: DescriptorVisibility,
            isVar: Boolean,
            source: SourceElement
        ): VariableDescriptorImpl {
            return VariableDescriptorImpl(containingDeclaration, name, null, isVar, source, visibility)
        }
    }


}
